### PR TITLE
AK/URL+URLParser: Implement the IPv4 number parser & host serializer

### PR DESCRIPTION
**Pull request by tgsm**

### CODE REVIEW

File: AK/URLParser.cpp [hunk 1]
```diff
@@ -44,6 +45,69 @@ static Optional<DeprecatedString> parse_opaque_host(StringView input)
     return URL::percent_encode(input, URL::PercentEncodeSet::C0Control);
 }
 
+// https://url.spec.whatwg.org/#ipv4-number-parser
+static Optional<Tuple<size_t, bool>> parse_ipv4_number(StringView input)
```
Comment (alimpfard) [inline AK/URLParser.cpp:50]: Please don't use Tuple like this.

File: AK/URLParser.cpp [hunk 2]
```diff
@@ -44,6 +45,69 @@ static Optional<DeprecatedString> parse_opaque_host(StringView input)
...
+    if (radix == 8)
+        output = AK::StringUtils::convert_to_uint_from_octal(input);
+    else if (radix == 10)
+        output = input.to_uint();
+    else if (radix == 16)
+        output = AK::StringUtils::convert_to_uint_from_hex(input);
+    VERIFY(output.has_value());
+
+    // 9. Return (output, validationError).
+    return Tuple<size_t, bool> { static_cast<size_t>(*output), validation_error };
```
Comment (alimpfard) [inline AK/URLParser.cpp:109]: Why the conversion to `size_t`? An ipv4 is always 32 bits

File: AK/URLParser.cpp [hunk 3]
```diff
@@ -110,8 +111,79 @@ static Optional<Tuple<size_t, bool>> parse_ipv4_number(StringView input)
 
 static Optional<DeprecatedString> parse_ipv4_address(StringView input)
```
Comment (alimpfard) [inline AK/URLParser.cpp:112]: Spec link while you're at it?

File: AK/URLParser.cpp [hunk 4]
```diff
@@ -110,8 +111,79 @@ static Optional<Tuple<size_t, bool>> parse_ipv4_number(StringView input)
...
+    }
+
+    // 3. If parts’s size is greater than 4, IPv4-too-many-parts validation error, return failure.
+    if (parts.size() > 4) {
+        report_validation_error();
+        return OptionalNone {};
+    }
+
+    // 4. Let numbers be an empty list.
+    Vector<size_t> numbers;
```
Comment (alimpfard) [inline AK/URLParser.cpp:134]: Please give this an inline capacity of 4.

File: AK/URLParser.cpp [hunk 5]
```diff
@@ -110,8 +111,79 @@ static Optional<Tuple<size_t, bool>> parse_ipv4_number(StringView input)
...
+    for (auto number : numbers) {
+        // 1. Increment ipv4 by n × 256^(3 − counter).
+        ipv4 += number * AK::pow<size_t>(256, 3 - counter);
+
+        // 2. Increment counter by 1.
+        counter++;
+    }
+
+    // 13. Return ipv4.
+    return DeprecatedString::number(ipv4);
```
Comment (alimpfard) [inline AK/URLParser.cpp:186]: Wait, why did this turn into a string?

**Comment by tgsm** [comment]
Superseded by #20176.

**Pull request closed without merging**